src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,917 @@
+
+                if StrLen(Buffer) > StrLen(Result) then
+                    Result := Buffer;
+                Buffer := '';
+                j += 1;
+            end;
+            i += 1;
+        end;
+
+        exit(Result);
+    end;
+
+
+    //TODO: This is a copy of codeunit 1251 "Record Match Mgt." GetLengthOfShortestString(). It should be moved to a system app, or replaced with a system function
+    local procedure GetLengthOfShortestString(FirstString: Text; SecondString: Text): Integer
+    begin
+        exit((StrLen(FirstString) + StrLen(SecondString) - Abs(StrLen(FirstString) - StrLen(SecondString))) / 2);
+    end;
+
+    //TODO: This is a copy of codeunit 1251 "Record Match Mgt." IsSubstringConsideredForNearness(). It should be moved to a system app, or replaced with a system function
+    local procedure IsSubstringConsideredForNearness(Substring: Text; MinThreshold: Integer): Boolean
+    var
+        Length: Integer;
+    begin
+        Length := StrLen(Substring);
+        if Length <= 1 then
+            exit(false);
+
+        exit(MinThreshold <= Length);
+    end;
+
+    local procedure RequiredNearness(): Decimal
+    begin
+        exit(0.9)
+    end;
+
+    local procedure ListAllTablesWithNumberSeries(var TablesPromptList: List of [Text])
+    var
+        TableMetadata: Record "Table Metadata";
+    begin
+        // Looping trhough all Setup tables
+        SetFilterOnSetupTables(TableMetadata);
+        if TableMetadata.FindSet() then
+            repeat
+                ListAllNoSeriesFields(TablesPromptList, TableMetadata);
+            until TableMetadata.Next() = 0;
+    end;
+
+    local procedure SetFilterOnSetupTables(var TableMetadata: Record "Table Metadata")
+    begin
+        TableMetadata.SetFilter(Name, '* Setup');
+        TableMetadata.SetRange(ObsoleteState, TableMetadata.ObsoleteState::No); //TODO: Check if 'Pending' should be included
+        TableMetadata.SetRange(TableType, TableMetadata.TableType::Normal);
+    end;
+
+    local procedure ListAllNoSeriesFields(var TablesPromptList: List of [Text]; var TableMetadata: Record "Table Metadata")
+    var
+        Field: Record "Field";
+    begin
+        SetFilterOnNoSeriesFields(TableMetadata, Field);
+        if Field.FindSet() then
+            repeat
+                AddNoSeriesFieldToTablesPrompt(TablesPromptList, TableMetadata, Field);
+            until Field.Next() = 0;
+    end;
+
+    local procedure AddNoSeriesFieldToTablesPrompt(var TablesPromptList: List of [Text]; TableMetadata: Record "Table Metadata"; Field: Record "Field")
+    begin
+        TablesPromptList.Add('Area: ' + RemoveTextPart(TableMetadata.Caption, ' Setup') + ', TableId: ' + Format(TableMetadata.ID) + ', FieldId: ' + Format(Field."No.") + ', FieldName: ' + RemoveTextPart(Field.FieldName, ' Nos.'));
+    end;
+
+    local procedure SetFilterOnNoSeriesFields(var TableMetadata: Record "Table Metadata"; var Field: Record "Field")
```

Comment:
I'm not a fan of this way of finding fields which set up No. Series. While it's a best practice to call those fields something with Nos., I don't think we can/should rely on that. Maybe we can use the following filter in some way?
`Field.SetRange(RelationTableNo, Database::"No. Series");`
Now you would filter on all fields which have a relationship to No. Series. Given that you filter on Setup tables, that would be setup of No. Series, I assume.

I'm not super happy with the "SetFilterOnSetupTables" procedure though, as again we probably cannot rely on the fact that all setup tables end on Setup. But I haven't found a solution to that yet :-)

Reply:
Especially there are affixes.
We mostly use prefix for our objects, but I know other partners that are using suffixes and therefore those setup tables won't be found.

Normally setup tables should have a primary key with Code[10].

Having an event to subscribe could be a solution but I'd prefer if we find some logic instead.